ext/dom/xpath.c

```diff
@@ -446,6 +446,60 @@ PHP_METHOD(DOMXPath, registerPhpFunctionNS)
 	);
 }
 
+/* {{{ */
+PHP_METHOD(DOMXPath, quote) {
+	char *input;
+	size_t input_len;
+	smart_str output = {0};
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &input, &input_len) ==
+			FAILURE) {
+		RETURN_THROWS();
+	}
+	if (memchr(input, '\'', input_len) == NULL) {
+		smart_str_appendc(&output, '\'');
+		smart_str_appendl(&output, input, input_len);
+		smart_str_appendc(&output, '\'');
+	} else if (memchr(input, '"', input_len) == NULL) {
+		smart_str_appendc(&output, '"');
+		smart_str_appendl(&output, input, input_len);
+		smart_str_appendc(&output, '"');
+	} else {
+		// need to use the concat() trick published by Robert Rossney at https://stackoverflow.com/a/1352556/1067003
+		smart_str_appends(&output, "concat(");
+		for (size_t i = 0; i < input_len;) {
+			uintptr_t bytesUntilSingleQuote =
+					(uintptr_t)memchr(input + i, '\'', input_len - i);
+			if (bytesUntilSingleQuote == 0) {
+				bytesUntilSingleQuote = input_len - i;
+			} else {
+				bytesUntilSingleQuote = bytesUntilSingleQuote - (uintptr_t)(input + i);
+			}
+			uintptr_t bytesUntilDoubleQuote =
+					(uintptr_t)memchr(input + i, '"', input_len - i);
+			if (bytesUntilDoubleQuote == 0) {
+				bytesUntilDoubleQuote = input_len - i;
+			} else {
+				bytesUntilDoubleQuote = bytesUntilDoubleQuote - (uintptr_t)(input + i);
+			}
+			const size_t bytesUntilQuote =
+					(bytesUntilSingleQuote > bytesUntilDoubleQuote)
+							? bytesUntilSingleQuote
+							: bytesUntilDoubleQuote;
+			const char quoteMethod =
+					(bytesUntilSingleQuote > bytesUntilDoubleQuote) ? '\'' : '"';
+			smart_str_appendc(&output, quoteMethod);
+			smart_str_appendl(&output, input + i, bytesUntilQuote);
+			smart_str_appendc(&output, quoteMethod);
+			i += bytesUntilQuote;
+			smart_str_appendc(&output, ',');
+		}
+		output.s->val[output.s->len -1 ] = ')'; // is there a smart_str function for this? (probably not)
```

Comment:
Indeed there is no way to do this.
```suggestion
		output.s->val[output.s->len - 1] = ')';
```